File: memwatch/bindings/MemWatchTest.cs
using System;
using System.Collections.Generic;

class MemWatchTest {
    class MemoryEvent {
        public string Name { get; set; }
        public int Offset { get; set; }
        public int OldValue { get; set; }
        public int NewValue { get; set; }
        
        public MemoryEvent(string name, int offset, int oldValue, int newValue) {
            Name = name;
            Offset = offset;
            OldValue = oldValue;
            NewValue = newValue;
        }
    }
    
    class MemoryTracker {
        Dictionary<int, byte[]> regions = new Dictionary<int, byte[]>();
        Dictionary<int, byte[]> initial = new Dictionary<int, byte[]>();
        List<MemoryEvent> events = new List<MemoryEvent>();
        int regionCounter = 0;
        
        public int Watch(byte[] data, string name) {
            int id = regionCounter++;
            regions[id] = (byte[])data.Clone();
            initial[id] = (byte[])data.Clone();
            Console.WriteLine($"  ✓ Watching region {id}: {name}");
            return id;
        }
        
        public void DetectChanges() {
            foreach (var kvp in regions) {
                int id = kvp.Key;
                byte[] region = kvp.Value;
                byte[] init = initial[id];
                
                for (int i = 0; i < region.Length; i++) {
                    if (init[i] != region[i]) {
                        events.Add(new MemoryEvent($"region_{id}", i, init[i], region[i]));
                        init[i] = region[i];
                    }
                }
            }
        }
        
        public int EventCount => events.Count;
        public List<MemoryEvent> Events => events;
    }
    
    static void Main() {
        Console.WriteLine("🧪 C# Memory Tracking Test");
        Console.WriteLine("==========================");
        
        MemoryTracker tracker = new MemoryTracker();
        byte[] data = new byte[20];
        Array.Fill(data, (byte)0);
        
        int regionId = tracker.Watch(data, "test_buffer");
        
        data[0] = 42;
        data[5] = 99;
        data[10] = 255;
        
        tracker.DetectChanges();
        
        if (tracker.EventCount == 3) {
            Console.WriteLine($"✅ PASS - C#: Detected {tracker.EventCount} changes");
            Console.WriteLine("   Events:");
            foreach (var evt in tracker.Events) {
                Console.WriteLine($"     - {evt.Name}[{evt.Offset}]: {evt.OldValue} -> {evt.NewValue}");
            }
            Environment.Exit(0);
        } else {
            Console.WriteLine($"❌ FAIL - C#: Expected 3 changes, got {tracker.EventCount}");
            Environment.Exit(1);
        }
    }
}

File: memwatch/bindings/MemWatchTest.java
/**
 * MemWatch Java Test
 * Tests basic Java memory tracking simulation
 */

import java.util.*;

public class MemWatchTest {
    static class MemoryEvent {
        String name;
        int offset;
        int oldValue;
        int newValue;
        
        MemoryEvent(String name, int offset, int oldValue, int newValue) {
            this.name = name;
            this.offset = offset;
            this.oldValue = oldValue;
            this.newValue = newValue;
        }
    }
    
    static class MemoryTracker {
        Map<Integer, byte[]> regions = new HashMap<>();
        Map<Integer, byte[]> initial = new HashMap<>();
        List<MemoryEvent> events = new ArrayList<>();
        int regionCounter = 0;
        
        int watch(byte[] data, String name) {
            int id = regionCounter++;
            byte[] dataCopy = new byte[data.length];
            System.arraycopy(data, 0, dataCopy, 0, data.length);
            
            byte[] initialCopy = new byte[data.length];
            System.arraycopy(data, 0, initialCopy, 0, data.length);
            
            regions.put(id, dataCopy);
            initial.put(id, initialCopy);
            System.out.println("  ✓ Watching region " + id + ": " + name);
            return id;
        }
        
        void detectChanges() {
            for (Integer id : regions.keySet()) {
                byte[] region = regions.get(id);
                byte[] init = initial.get(id);
                for (int i = 0; i < region.length; i++) {
                    if (init[i] != region[i]) {
                        events.add(new MemoryEvent("region_" + id, i, init[i] & 0xFF, region[i] & 0xFF));
                        init[i] = region[i];
                    }
                }
            }
        }
    }
    
    public static void main(String[] args) {
        System.out.println("🧪 Java Memory Tracking Test");
        System.out.println("============================");
        
        MemoryTracker tracker = new MemoryTracker();
        byte[] data = new byte[20];
        Arrays.fill(data, (byte)0);
        
        int regionId = tracker.watch(data, "test_buffer");
        
        // Modify the array
        data[0] = 42;
        data[5] = 99;
        data[10] = (byte)255;
        
        // Update region to reflect changes
        tracker.regions.put(regionId, data);
        
        tracker.detectChanges();
        
        if (tracker.events.size() == 3) {
            System.out.println("✅ PASS - Java: Detected " + tracker.events.size() + " changes");
            System.out.println("   Events:");
            for (MemoryEvent evt : tracker.events) {
                System.out.println("     - " + evt.name + "[" + evt.offset + "]: " + evt.oldValue + " -> " + evt.newValue);
            }
            System.exit(0);
        } else {
            System.out.println("❌ FAIL - Java: Expected 3 changes, got " + tracker.events.size());
            System.exit(1);
        }
    }
}
